pjmedia/src/pjmedia/sound_port.c

```diff
@@ -322,24 +383,79 @@ static pj_status_t start_sound_device( pj_pool_t *pool,
         status = pjmedia_snd_port_set_ec(snd_port, pool, 
                                          snd_port->aud_param.ec_tail_ms,
                                          snd_port->prm_ec_options);
-        if (status != PJ_SUCCESS) {
-            pjmedia_aud_stream_destroy(snd_port->aud_stream);
-            snd_port->aud_stream = NULL;
-            return status;
+        if (status != PJ_SUCCESS)
+            goto on_error;
+    }
+
+    /* Create clock and buffers, if configured */
+    if ((snd_port->options & PJMEDIA_SND_PORT_USE_SOFT_CLOCK) &&
+        (snd_port->aud_param.ext_fmt.id == PJMEDIA_FORMAT_L16))
+    {
+        unsigned ptime;
+
+        status = pjmedia_clock_create(pool,
+                                      snd_port->clock_rate,
+                                      snd_port->channel_count,
+                                      snd_port->samples_per_frame,
+                                      0,
+                                      &clock_callback,
+                                      snd_port,
+                                      &snd_port->clock);
+        if (status != PJ_SUCCESS)
+            goto on_error;
+
+        ptime = snd_port->samples_per_frame * 1000 / snd_port->clock_rate /
+                snd_port->channel_count;
+        status = pjmedia_delay_buf_create(pool, "playdbuf",
+                                          snd_port->clock_rate,
+                                          snd_port->samples_per_frame,
+                                          snd_port->channel_count,
+                                          PJMEDIA_SOUND_BUFFER_COUNT * ptime,
+                                          0, /* options */
+                                          &snd_port->play_dbuf);
+        if (status != PJ_SUCCESS)
+            goto on_error;
+
+        status = pjmedia_delay_buf_create(pool, "capdbuf",
+                                          snd_port->clock_rate,
+                                          snd_port->samples_per_frame,
+                                          snd_port->channel_count,
+                                          PJMEDIA_SOUND_BUFFER_COUNT * ptime,
+                                          0, /* options */
+                                          &snd_port->cap_dbuf);
+        if (status != PJ_SUCCESS)
+            goto on_error;
+
+        snd_port->frame_buf = (pj_int16_t*)
+                              pj_pool_zalloc(pool,
+                                             snd_port->samples_per_frame * 2);
+        if (!snd_port->frame_buf) {
+            status = PJ_ENOMEM;
+            goto on_error;
         }
+
+        status = pjmedia_clock_start(snd_port->clock);
+        if (status != PJ_SUCCESS)
+            goto on_error;
+
+        PJ_LOG(4,(THIS_FILE, "Sound port does not use native clock"));
```

Comment:
Explicit would be better here, as to not leave the users confused (the users know it doesn't use native clock but then they don't know what's being used). So "Sound port uses internal (or software) clock" should be more clear.